pkg/rrd/rrd.go

```diff
@@ -0,0 +1,224 @@
+	// name
+	Last(key string) (value float64, ok bool, err error)
+}
+
+type Slot interface {
+	// Counter sets (or overrides) the current stored value for this key,
+	// with value
+	Counter(key string, value float64) error
+	// Key return the key of the slot which is the window timestamp
+	Key() uint64
+}
+
+type rrdBolt struct {
+	db        *bolt.DB
+	window    uint64
+	retention uint64
+}
+
+type rrdSlot struct {
+	db  *bolt.DB
+	key uint64
+}
+
+// NewRRDBolt creates a new rrd database that uses bolt as a storage. if window or retention are 0
+// the function will panic. If retnetion is smaller then window the function will panic.
+// retention and window must be multiple of 1 minute.
+func NewRRDBolt(path string, window time.Duration, retention time.Duration) (RRD, error) {
+	return newRRDBolt(path, window, retention)
+}
+
+func newRRDBolt(path string, window time.Duration, retention time.Duration) (*rrdBolt, error) {
+	window = (window / time.Minute) * time.Minute
+	retention = (retention / time.Minute) * time.Minute
+
+	if window == 0 {
+		panic("invalid window, can't be zero")
+	}
+	if retention == 0 {
+		panic("invalid retention, can't be zero")
+	}
+	if retention < window {
+		panic("retention can't be smaller than window")
+	}
+
+	db, err := bolt.Open(path, 0644, bolt.DefaultOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open bolt db")
+	}
+
+	return &rrdBolt{
+		db:        db,
+		window:    uint64(window / time.Second),
+		retention: uint64(retention / time.Second),
+	}, nil
+}
+
+func (r *rrdBolt) retain(now uint64) error {
+	retain := now - r.retention
+	return r.db.Update(func(tx *bolt.Tx) error {
+		cur := tx.Cursor()
+
+		for k, _ := cur.First(); k != nil; k, _ = cur.Next() {
+			if len(k) != 8 {
+				continue // uknown key size
+			}
+
+			if lu64(k) <= retain {
+				// delete the bucket
+				// are we sure this is safe to do while iterating ?
+				_ = tx.DeleteBucket(k)
+			}
+		}
+
+		return nil
+	})
+}
+
+func (r *rrdBolt) Slots() ([]uint64, error) {
+	var slots []uint64
+	err := r.db.View(func(tx *bolt.Tx) error {
+		cur := tx.Cursor()
+		for k, _ := cur.First(); k != nil; k, _ = cur.Next() {
+			slots = append(slots, lu64(k))
+		}
+
+		return nil
+	})
+
+	return slots, err
+}
+
+func (r *rrdBolt) Last(key string) (value float64, ok bool, err error) {
+	err = r.db.View(func(tx *bolt.Tx) error {
+		cur := tx.Cursor()
+		for k, _ := cur.Last(); k != nil; k, _ = cur.Prev() {
+			bucket := tx.Bucket(k)
+			bytes := bucket.Get([]byte(key))
+			if bytes != nil {
+				value = lf64(bytes)
+				ok = true
+				break
+			}
+		}
+
+		return nil
+	})
+
+	return
+}
+
+// Counters return increase in counter value since the given
+// start time.
+func (r *rrdBolt) Counters(since time.Time) (map[string]float64, error) {
+	ts := uint64(since.Unix())
+	ts = (ts / r.window) * r.window
+
+	// we start from the previous slot so we check from the last value.
+	ts -= r.window
```

Comment:
It's not very clear (to me) what increments `Counters` exactly considers. But subtracting the window after flooring to the last Slot will cause the increments inside the slot 70 to be counted twice when calculating `Counters(13) at 73 + Counters(73) at 133` when window is 5, since ts in the second call `ts` will end up equal to 65. Not subtracting the window will consider increments after since (not or at since) which will make consecutive calls consider distinct increments and not skip any slot.

Reply:
Yeah, may be "Counters" name is not very clear but this computes the "increase" on a specific metric over time. The idea is i take a sample of the reading (network card) every "window" (let's say 5 min) then it's
```
0     5   10  15  20
|-----|-----|----|----|
    ^     ^     ^   ^
```
Then I can compute how much (traffic) was on the card from minute 0 to minute 20, by adding the differences between values as (v5-v0) + (v10-v5) + ... etc.

The problem is now I don't know what is the increase from value at -5, to value at 0 
So to get the cosumption at slot 0-5 i need to know the value at -5, right ? 
hence i start from the previous slot from the 'since'